server/scsynth/SC_PortAudio.cpp

```diff
@@ -262,39 +270,181 @@ int SC_PortAudioDriver::PortAudioCallback(const void* input, void* output, unsig
-            if (pdi->maxInputChannels > 0)
-                mInOut[0] = i;
-            if (pdi->maxOutputChannels > 0)
-                mInOut[1] = i;
+        auto* pdi = Pa_GetDeviceInfo(i);
+        std::string devString = GetPaDeviceName(i);
+        if (strstr(devString.c_str(), device)) {
+            if (ioType == IOType::Input && pdi->maxInputChannels > 0) {
+                *mInOut = i;
+                break;
+            } else if (ioType == IOType::Output && pdi->maxOutputChannels > 0) {
+                *mInOut = i;
+                break;
+            }
         }
     }
 }
 
+PaStreamParameters SC_PortAudioDriver::GetPaStreamParameters(int device, int channelCount,
+                                                             double suggestedLatency) const {
+    PaStreamParameters streamParams;
+    PaSampleFormat fmt = paFloat32 | paNonInterleaved;
+    streamParams.device = device;
+    streamParams.channelCount = channelCount;
+    streamParams.sampleFormat = fmt;
+    streamParams.suggestedLatency = suggestedLatency;
+    streamParams.hostApiSpecificStreamInfo = nullptr;
+    return streamParams;
+}
+
+template <typename SupportCheck> // expected signature: PaError (*)(PaStreamParameters&, double)
+PaError SC_PortAudioDriver::CheckSinglePaDevice(int* device, double sampleRate, int maxChannels, int defaultDevice,
+                                                SupportCheck isSupportedFunc, const char* deviceType) const {
+    if (*device != paNoDevice && sampleRate) {
+        // check if device can support requested SR
+        PaStreamParameters parameters = GetPaStreamParameters(*device, maxChannels, 0);
+        PaError err = isSupportedFunc(parameters, sampleRate);
+        if (err != paNoError) {
+            fprintf(stdout, "PortAudio error: %s\nRequested sample rate %f for device %s is not supported\n",
+                    Pa_GetErrorText(err), sampleRate, Pa_GetDeviceInfo(*device)->name);
+            return err;
+        }
+    }
+    // in case we still don't have a proper device, use the default device
+    if (*device == paNoDevice) {
+        *device = defaultDevice;
+        if (*device != paNoDevice)
+            fprintf(stdout, "Selecting default system %s device\n", deviceType);
+    }
+    return paNoError;
+}
+
+void SC_PortAudioDriver::SelectMatchingPaDevice(int* matchingDevice, int* knownDevice,
+                                                IOType matchingDeviceType) const {
+    if (*matchingDevice != paNoDevice || *knownDevice == paNoDevice)
+        return;
+
+    const PaHostApiInfo* apiInfo;
+    apiInfo = Pa_GetHostApiInfo(Pa_GetDeviceInfo(*knownDevice)->hostApi);
+
+    bool isInput;
+    if (matchingDeviceType == IOType::Input)
+        isInput = true;
+    else if (matchingDeviceType == IOType::Output)
+        isInput = false;
+
+    if (isInput) {
+        *matchingDevice = apiInfo->defaultInputDevice;
+    } else {
+        *matchingDevice = apiInfo->defaultOutputDevice;
+    }
```

Comment:
grr, i've been going back and forth on the best way to handle this function. first of all, I think a more appropriate name would be something that indicates that the function may not do anything. "TryMatchUnspecifiedDevice"? "Try" would be a good word to have in the function name in any case. It also indicates that the function is encapsulating not just the matching, but the checks that happen prior to it.

`const int* knownDevice` in the parameter list would make it immediately much clearer which variable is the output and which is the input.

the way you set `isInput` is a little too careful -- we have control over all the uses of this function, it's easy to confirm that you will only ever have one or the other.

A more compact but still straightforward way to write this block would be:

```cpp
    auto* apiInfo = Pa_GetHostApiInfo(Pa_GetDeviceInfo(*knownDevice)->hostApi);
    bool isInput = matchingDeviceType == IOType::Input;
    *matchingDevice = isInput ? apiInfo->defaultInputDevice : apiInfo->defaultOutputDevice;
```

in fact, I would say this is even clearer -- the data flow is not interrupted by `if`s, and it's clear that the goal is to set `matchingDevice`. 3 lines instead of your current 14. up to you if you want to change it; though I'd at least recommend not using 5 lines to set `isInput` when 1 would do.